bionic/executor.py

```diff
@@ -6,19 +6,83 @@
 """
 
 import copy
+from functools import partial
 import logging
+from multiprocessing.managers import SyncManager
 import queue
 import sys
 import threading
 import traceback
+from uuid import uuid4
 
-from multiprocessing.managers import SyncManager
-
+from .aip.task import Task
 from .deps.optdep import import_optional_dependency
 from .utils.misc import oneline, SynchronizedSet
 
 
-class Executor:
+class _BaseExecutor:
+    def submit(self, fn, *args, **kwargs):
+        """
+        Submits the function to the executor and returns a Future object.
+        """
+        pass
+
+    def create_synchronized_set(self):
+        """
+        Creates a set that is used by TaskKeyLogger for progress reporting.
+        """
+        pass
+
+    def start_logging(self):
+        """
+        Sets up anything related to logging in the current process for logs to
+        work for the submitted tasks.
+        """
+        pass
+
+    def stop_logging(self):
+        """
+        Stops any logging work in the current process that is not needed anymore.
+        """
+        pass
+
+    def set_resource(self, resource):
+        """This is a workaround for AIP PoC."""
+        pass
+
+
+class AipExecutor(_BaseExecutor):
+    """
+    Encapsulates all objects related to remote execution on Google AIP
+    in one place.
+    """
+
+    def __init__(self, aip_job):
+        self._aip_job = aip_job
+
+    def submit(self, fn, *args, **kwargs):
+        assert self._resource is not None
+        return Task(
+            # TODO: Use a better identifiable name, maybe a combination
+            # of entity name and case key.
+            # This is a temporary workaround. We are changing the random
+            # UUID in such a manner because AIP names have to start with
+            # an alphabet and only accepts alphanumeric and underscore
+            # characters.
+            name="a" + str(uuid4()).replace("-", ""),
+            job=self._aip_job,
+            resource=self._resource,
+            function=partial(fn, *args, **kwargs),
+        ).submit()
+
+    def create_synchronized_set(self):
+        return SynchronizedSet()
```

Comment:
Does this work on AIP?

Reply:
Depends on what you mean by "work". AIP would get the synchronized set, but the main process that's sending the remote jobs won't get back the updates on the set. As we discussed before, progress reporting would be broken with this PoC. 

I can add this as a code comment too.